File: leetcode/lc_067.c
//
// LeetCode
// 67. Add Binary
// https://leetcode.com/problems/add-binary/
//


char* addBinary(char* a, char* b) {
    char* result;
    int size, i, j, n;
    
    i = strlen(a);
    j = strlen(b);
    size = (i > j ? i : j) + 2;
    
    if ((result = (char *) malloc(size * sizeof(char))) == NULL)
        exit(-1);
    
    result[--size] = '\0';
    i--;
    j--;
    size--;
    n = 0;
    while (size > 0) {
        if (i >= 0 && j >= 0)
            n += (a[i--] - '0') + (b[j--] - '0');
        else if (i >= 0)
            n += a[i--] - '0';
        else if (j >= 0)
            n += b[j--] - '0';
        
        result[size--] = n % 2 + '0';
        
        n = n / 2;
    }
    
    if (n > 0)
        result[size] = n % 2 + '0';
    else
        result++;
    
    return result;
}

File: leetcode-swift/lc_067.swift
//
// LeetCode
// 67. Add Binary
// https://leetcode.com/problems/add-binary/
//


class Solution {
    func addBinary(_ a: String, _ b: String) -> String {
        var result: String = ""
        var carry = 0
        var i = a.count - 1
        var j = b.count - 1
        
        while i >= 0 || j >= 0 {
            var sum = carry
            
            if i >= 0 {
                let aIndex = a.index(a.startIndex, offsetBy: i)
                sum += Int(String(a[aIndex]))!
            }
            
            if j >= 0 {
                let bIndex = b.index(b.startIndex, offsetBy: j)
                sum += Int(String(b[bIndex]))!
            }
            
            result = String(sum % 2) + result
            carry = sum / 2

            i -= 1
            j -= 1
        }
        
        if carry > 0 {
            result = String(carry) + result
        }
        
        return result
    }
}
